### PR TITLE
fix sendTo and recvFrom in asyncnet

### PR DESCRIPTION
**Pull request by rockcavera**
* As discussed in pull [#14109](https://github.com/nim-lang/Nim/pull/14109), the pointers on parameters of the high-level procs were removed.
* Now there are only 2 procs. Before there were 4.

Open to suggestions.

### CODE REVIEW

File: lib/pure/asyncnet.nim [hunk 1]
```diff
@@ -780,9 +780,9 @@ proc isClosed*(socket: AsyncSocket): bool =
   ## Determines whether the socket has been closed.
   return socket.closed
 
-proc sendTo*(socket: AsyncSocket, data: pointer, dataSize: int, address: string,
-             port: Port, flags = {SocketFlag.SafeDisconn}): owned(Future[void])
-            {.async, since: (1, 3).} =
+proc sendTo*(socket: AsyncSocket, data: string, address: string, port: Port,
+             flags = {SocketFlag.SafeDisconn}): owned(Future[void])
+            {.async, since: (1, 3, 1).} =
```
Comment (dom96) [inline lib/pure/asyncnet.nim]: Pretty sure this `, 1` isn't necessary.

File: lib/pure/asyncnet.nim [hunk 2]
```diff
@@ -780,9 +780,9 @@ proc isClosed*(socket: AsyncSocket): bool =
   ## Determines whether the socket has been closed.
   return socket.closed
 
-proc sendTo*(socket: AsyncSocket, data: pointer, dataSize: int, address: string,
-             port: Port, flags = {SocketFlag.SafeDisconn}): owned(Future[void])
-            {.async, since: (1, 3).} =
+proc sendTo*(socket: AsyncSocket, data: string, address: string, port: Port,
```
Comment (dom96) [inline lib/pure/asyncnet.nim]: Please change order of arguments so that it matches the ones in `net`: https://nim-lang.org/docs/net.html#sendTo%2CSocket%2Cstring%2CPort%2Cstring

File: lib/pure/asyncnet.nim [hunk 3]
```diff
@@ -824,88 +822,58 @@ proc sendTo*(socket: AsyncSocket, data: pointer, dataSize: int, address: string,
...
-  ## This proc sends ``data`` to the specified ``address``, which may be an IP
-  ## address or a hostname. If a hostname is specified this function will try
-  ## each IP of that hostname. The returned future will complete once all data
-  ## has been sent.
+      raise newException(IOError, "Couldn't resolve address: " & address)
+
+proc recvFrom*(socket: AsyncSocket, size: int,
+               flags = {SocketFlag.SafeDisconn}):
+              owned(Future[tuple[data: string, address: string, port: Port]])
+              {.async, since: (1, 3, 1).} =
```
Comment (dom96) [inline lib/pure/asyncnet.nim]: This is fine, but if you're up for it, a `recvFrom` taking a `FutureVar` and matching the net.recvFrom proc (https://nim-lang.org/docs/net.html#recvFrom%2CSocket%2Cstring%2Cint%2Cstring%2CPort%2Cint32) would be nice. Might be better to create it in a separate PR.

File: lib/pure/asyncnet.nim [hunk 4]
```diff
@@ -824,88 +822,58 @@ proc sendTo*(socket: AsyncSocket, data: pointer, dataSize: int, address: string,
...
-
-    awaitRecvFromInto()
+    yield fut
 
-    port.complete(ntohs(sAddr.sin6_port).Port)
-    
-  else:
-    var sAddr: Sockaddr_in
+    if fut.failed():
+      raise fut.readError()
```
Comment (dom96) [inline lib/pure/asyncnet.nim]: Can you use ``await`` here instead, it has the same semantics (exceptions are re-thrown).

**Review by dom96 — APPROVED** [review]
Overal looks good, but please fix the `sendTo` API.

File: lib/pure/asyncnet.nim (hunk 4 above)
Comment (rockcavera) [reply lib/pure/asyncnet.nim]: It's done.

File: lib/pure/asyncnet.nim (hunk 2 above)
Comment (rockcavera) [reply lib/pure/asyncnet.nim]: It's done.

File: lib/pure/asyncnet.nim (hunk 1 above)
Comment (rockcavera) [reply lib/pure/asyncnet.nim]: It's done.

File: lib/pure/asyncnet.nim (hunk 3 above)
Comment (rockcavera) [reply lib/pure/asyncnet.nim]: Let me see if I understand, because my English is not the best.

Create another proc recvFrom (using net.recvFrom as a model), but using FutureVars in the parameters? In that, would be two recvFrom in asyncnet?

Comment (dom96) [reply lib/pure/asyncnet.nim]: yep, that's correct. You can do this in a separate PR.
